consumers/writers/api/logging.go

```diff
@@ -13,30 +13,65 @@ import (
 	log "github.com/mainflux/mainflux/logger"
 )
 
-var _ consumers.Consumer = (*loggingMiddleware)(nil)
+var _ consumers.AsyncConsumer = (*asyncLoggingMiddleware)(nil)
+var _ consumers.SyncConsumer = (*syncLoggingMiddleware)(nil)
 
-type loggingMiddleware struct {
-	logger   log.Logger
-	consumer consumers.Consumer
+type asyncLoggingMiddleware struct {
+	logger        log.Logger
+	asyncConsumer consumers.AsyncConsumer
 }
 
-// LoggingMiddleware adds logging facilities to the adapter.
-func LoggingMiddleware(consumer consumers.Consumer, logger log.Logger) consumers.Consumer {
-	return &loggingMiddleware{
-		logger:   logger,
-		consumer: consumer,
+// AsyncLoggingMiddleware adds logging facilities to the adapter.
+func AsyncLoggingMiddleware(asyncConsumer consumers.AsyncConsumer, logger log.Logger) consumers.AsyncConsumer {
+	return &asyncLoggingMiddleware{
+		logger:        logger,
+		asyncConsumer: asyncConsumer,
 	}
 }
 
-func (lm *loggingMiddleware) Consume(msgs interface{}) (err error) {
+func (alm *asyncLoggingMiddleware) ConsumeAsync(msgs interface{}) {
+	ch := alm.asyncConsumer.Errors()
+	var err error
+
+	defer func(begin time.Time) {
+		message := fmt.Sprintf("Method ConsumeAsync took %s to complete", time.Since(begin))
+		if err != nil {
+			alm.logger.Warn(fmt.Sprintf("%s with error: %s.", message, err))
+			return
+		}
+		alm.logger.Info(fmt.Sprintf("%s without errors.", message))
+	}(time.Now())
+
+	go alm.asyncConsumer.ConsumeAsync(msgs)
+	err = <-ch
```

Comment:
Not only this is incorrect because errors need to be processed **in an infinite loop**, but this will also cause **consumer to block if the errors channel is unbuffered**. I'd suggest that you remove AsyncLogger (and handle logging in the `main.go`) because, unlike for sync operations, the error consumer is an integral part of the async consumers and is not only the logger but also the error handler.